### ISSUE TITLE
Reschedule `jaeger-cassandra-schema-job` upon failure

### ISSUE THREAD

**Issue by nlamirault** [jaegertracing/jaeger-kubernetes#32]
I'm using the production YAML file to deploy Jager : 

```
root@kmaster01:~# kubectl  --namespace=opentracing get pod
NAME                                READY     STATUS             RESTARTS   AGE
cassandra-0                         1/1       Running            0          5m
cassandra-1                         1/1       Running            1          5m
cassandra-2                         1/1       Running            3          5m
jaeger-collector-1578691591-gm9dn   0/1       CrashLoopBackOff   3          6m
jaeger-query-1666845303-113hb       0/1       CrashLoopBackOff   2          6m

root@kmaster01:~# kubectl  --namespace=opentracing logs jaeger-collector-1578691591-gm9dn
2017/09/06 14:42:56 error: failed to connect to 10.244.48.54:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
2017/09/06 14:42:56 error: failed to connect to 10.244.80.25:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
2017/09/06 14:42:56 error: failed to connect to 10.244.128.36:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
2017/09/06 14:42:56 error: failed to connect to 10.244.48.54:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
2017/09/06 14:42:56 error: failed to connect to 10.244.80.25:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
2017/09/06 14:42:57 error: failed to connect to 10.244.128.36:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
{"level":"fatal","ts":1504708977.0914874,"caller":"collector/main.go:79","msg":"Unable to set up builder","error":"no connections were made when creating the session","stacktrace":"github.com/uber/jaeger/vendor/go.uber.org/zap.Stack\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/go.uber.org/zap/field.go:209\ngithub.com/uber/jaeger/vendor/go.uber.org/zap.(*Logger).check\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/go.uber.org/zap/logger.go:273\ngithub.com/uber/jaeger/vendor/go.uber.org/zap.(*Logger).Fatal\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/go.uber.org/zap/logger.go:208\nmain.main.func1\n\t/home/travis/gopath/src/github.com/uber/jaeger/cmd/collector/main.go:79\ngithub.com/uber/jaeger/vendor/github.com/spf13/cobra.(*Command).execute\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/github.com/spf13/cobra/command.go:636\ngithub.com/uber/jaeger/vendor/github.com/spf13/cobra.(*Command).ExecuteC\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/github.com/spf13/cobra/command.go:722\ngithub.com/uber/jaeger/vendor/github.com/spf13/cobra.(*Command).Execute\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/github.com/spf13/cobra/command.go:681\nmain.main\n\t/home/travis/gopath/src/github.com/uber/jaeger/cmd/collector/main.go:122"}

root@kmaster01:~# kubectl  --namespace=opentracing logs jaeger-query-1666845303-113hb
2017/09/06 14:43:12 error: failed to connect to 10.244.128.36:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
2017/09/06 14:43:12 error: failed to connect to 10.244.80.25:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
2017/09/06 14:43:12 error: failed to connect to 10.244.48.54:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
2017/09/06 14:43:12 error: failed to connect to 10.244.128.36:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
2017/09/06 14:43:12 error: failed to connect to 10.244.80.25:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
2017/09/06 14:43:12 error: failed to connect to 10.244.48.54:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
{"level":"fatal","ts":1504708993.0407972,"caller":"query/main.go:71","msg":"Failed to init storage builder","error":"no connections were made when creating the session","stacktrace":"github.com/uber/jaeger/vendor/go.uber.org/zap.Stack\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/go.uber.org/zap/field.go:209\ngithub.com/uber/jaeger/vendor/go.uber.org/zap.(*Logger).check\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/go.uber.org/zap/logger.go:273\ngithub.com/uber/jaeger/vendor/go.uber.org/zap.(*Logger).Fatal\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/go.uber.org/zap/logger.go:208\nmain.main.func1\n\t/home/travis/gopath/src/github.com/uber/jaeger/cmd/query/main.go:71\ngithub.com/uber/jaeger/vendor/github.com/spf13/cobra.(*Command).execute\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/github.com/spf13/cobra/command.go:636\ngithub.com/uber/jaeger/vendor/github.com/spf13/cobra.(*Command).ExecuteC\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/github.com/spf13/cobra/command.go:722\ngithub.com/uber/jaeger/vendor/github.com/spf13/cobra.(*Command).Execute\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/github.com/spf13/cobra/command.go:681\nmain.main\n\t/home/travis/gopath/src/github.com/uber/jaeger/cmd/query/main.go:100"}
```

Any idea why i've got these errors ? Thanks.

**Comment by jpkrohling** [comment]
TL;DR: For the first couple of minutes or so, it should be fine to have those errors. Kubernetes will eventually spin up new instances of the pods and it will eventually work.

Long version: When deploying everything, we can't[1] have an orderly deployment order. Basically, we need to deploy Cassandra, create the schema required by Jaeger, deploy the Collector and the Query services. What you are seeing is the result of the Collector/Query starting before the schema was created, and after Cassandra has been initialized. The "create schema" is a Kubernetes job (`kubectl logs job/jaeger-cassandra-schema-job`) which waits up until 1 minute for Cassandra to be ready. If it times out, Kubernetes will schedule a new run (IIRC, not 100% sure on that now).

1 - I *think* there's a new feature on newer versions of Kubernetes where we could add the create-schema job, but I need to check that

**Comment by jpkrohling** [comment]
By the way: if those won't go away after a few minutes, please do post the logs for the create-schema job:

```bash
$ kubectl logs job/jaeger-cassandra-schema-job
```

**Comment by jpkrohling** [comment]
By the way: if those won't go away after a few minutes, please do post the logs for the create-schema job:

```bash
$ kubectl logs job/jaeger-cassandra-schema-job
```

**Comment by nlamirault** [comment]
@jpkrohling ok thanks

```
root@kmaster01:~# kubectl --namespace=opentracing get job
NAME                          DESIRED   SUCCESSFUL   AGE
jaeger-cassandra-schema-job   1         0            26m
root@kmaster01:~# kubectl --namespace=opentracing logs job/jaeger-cassandra-schema-job
Found 10 pods, using pod/jaeger-cassandra-schema-job-ggw25
Error from server (BadRequest): container "jaeger-cassandra-schema" in pod "jaeger-cassandra-schema-job-ggw25" is not available
```

I will wait a few minutes.

**Comment by jpkrohling** [comment]
26 minutes is way too long already :) It looks like I'm wrong and Kubernetes won't reschedule a new run for the failed create-schema job. Could you try to remove it and deploy the services again? Something like:

```
kubectl delete job/jaeger-cassandra-schema-job
kubectl create -f https://raw.githubusercontent.com/jaegertracing/jaeger-kubernetes/master/production/jaeger-production-template.yml
```

(the last command should fail on all services except the create-schema job)

**Comment by nlamirault** [comment]
Output : 

```
job "jaeger-cassandra-schema-job" created
Error from server (AlreadyExists): services "cassandra" already exists
Error from server (AlreadyExists): statefulsets.apps "cassandra" already exists
Error from server (AlreadyExists): deployments.extensions "jaeger-collector" already exists
Error from server (AlreadyExists): services "jaeger-collector" already exists
Error from server (AlreadyExists): services "zipkin" already exists
Error from server (AlreadyExists): deployments.extensions "jaeger-query" already exists
Error from server (AlreadyExists): services "jaeger-query" already exists
root@kmaster01:~# kubectl --namespace=opentracing logs job/jaeger-cassandra-schema-job

system_traces  system_schema  system_auth  system  system_distributed

Generating the schema for the keyspace jaeger_v1_dc1 and datacenter dc1
Using template file /cassandra-schema/v001.cql.tmpl with parameters:
    mode = prod
    datacenter = dc1
    keyspace = jaeger_v1_dc1
    replication = {'class': 'NetworkTopologyStrategy', 'dc1': '2' }
    trace_ttl = 172800
    dependencies_ttl = 0

```

**Comment by jpkrohling** [comment]
Looks like your collector and query are now working. 

The issue is then to make sure the create-schema job can be rescheduled upon failure.

**Comment by heppu** [comment]
I had the same problem but deleting and recreating the job fixed it.

**Comment by jpkrohling** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Reschedule `jaeger-cassandra-schema-job` upon failure](https://trello.com/c/nUDIVIIf/79-reschedule-jaeger-cassandra-schema-job-upon-failure)

**Comment by jpkrohling** [comment]
@heppu , @nlamirault , would you be able to give the fix a try and confirm it works?

**Comment by bigwhite** [comment]
I have met the same problem. The version of jaeger I use is below:
```
root@ubuntu-k8s-3:~# docker images|grep jaeger
jaegertracing/jaeger-query                                 0.9                 2e85f0280681        11 days ago         41.9 MB
jaegertracing/jaeger-collector                             0.9                 ab916122f094        11 days ago         20 MB
jaegertracing/jaeger-cassandra-schema                      0.6                 27f8532f0ab3        2 months ago        386 MB
root@ubuntu-k8s-3:~# docker images|grep cassandra
cassandra                                                  3.11                b2a0d42e1e97        2 days ago          385 MB
jaegertracing/jaeger-cassandra-schema                      0.6                 27f8532f0ab3        2 months ago        386 MB

``

First I deploy cassandra pod with production/cassandra.yml. 10 min later  I use jaeger-production-template.yml to deploy jaeger backend.

Then I found jaeger backends pods are all not working:

```
default                    jaeger-collector-216026522-v49hz       0/1       CrashLoopBackOff   4          2m        10.46.0.8      ubuntu-k8s-3
default                    jaeger-query-2386461091-wsh3g          0/1       CrashLoopBackOff   4          2m        10.46.0.9      ubuntu-k8s-3
```

The log of collector and query are below:

```
root@ubuntu-k8s-3:~# docker logs 1554b653aa0f
{"level":"info","ts":1510020066.1924856,"caller":"healthcheck/handler.go:46","msg":"Health Check server started","http-port":14269}
2017/11/07 02:01:06 error: failed to connect to 10.46.0.7:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
2017/11/07 02:01:06 error: failed to connect to 10.46.0.7:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
{"level":"fatal","ts":1510020066.3941922,"caller":"collector/main.go:88","msg":"Unable to set up builder","error":"no connections were made when creating the session","stacktrace":"main.main.func1\n\t/home/travis/gopath/src/github.com/uber/jaeger/cmd/collector/main.go:88\ngithub.com/uber/jaeger/vendor/github.com/spf13/cobra.(*Command).execute\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/github.com/spf13/cobra/command.go:636\ngithub.com/uber/jaeger/vendor/github.com/spf13/cobra.(*Command).ExecuteC\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/github.com/spf13/cobra/command.go:722\ngithub.com/uber/jaeger/vendor/github.com/spf13/cobra.(*Command).Execute\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/github.com/spf13/cobra/command.go:681\nmain.main\n\t/home/travis/gopath/src/github.com/uber/jaeger/cmd/collector/main.go:142\nruntime.main\n\t/home/travis/.gimme/versions/go1.9.linux.amd64/src/runtime/proc.go:185"}

root@ubuntu-k8s-3:~# docker logs 1e791f7ac916
{"level":"info","ts":1510020071.1615207,"caller":"healthcheck/handler.go:46","msg":"Health Check server started","http-port":16687}
2017/11/07 02:01:11 error: failed to connect to 10.46.0.7:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
2017/11/07 02:01:11 error: failed to connect to 10.46.0.7:9042 due to error: Keyspace 'jaeger_v1_dc1' does not exist
{"level":"fatal","ts":1510020071.37031,"caller":"query/main.go:92","msg":"Failed to init storage builder","error":"no connections were made when creating the session","stacktrace":"main.main.func1\n\t/home/travis/gopath/src/github.com/uber/jaeger/cmd/query/main.go:92\ngithub.com/uber/jaeger/vendor/github.com/spf13/cobra.(*Command).execute\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/github.com/spf13/cobra/command.go:636\ngithub.com/uber/jaeger/vendor/github.com/spf13/cobra.(*Command).ExecuteC\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/github.com/spf13/cobra/command.go:722\ngithub.com/uber/jaeger/vendor/github.com/spf13/cobra.(*Command).Execute\n\t/home/travis/gopath/src/github.com/uber/jaeger/vendor/github.com/spf13/cobra/command.go:681\nmain.main\n\t/home/travis/gopath/src/github.com/uber/jaeger/cmd/query/main.go:135\nruntime.main\n\t/home/travis/.gimme/versions/go1.9.linux.amd64/src/runtime/proc.go:185"}
```

According the advice of @jpkrohling,  I got the log of jaeger-cassandra-schema-job and it seems the job work unexpectedly:

```
# kubectl logs job/jaeger-cassandra-schema-job
error: timed out waiting for the condition
```

I delete the job and recreate it. The restarting log is below:

```
# kubectl logs  job/jaeger-cassandra-schema-job

system_traces  system_schema  system_auth  system  system_distributed

Generating the schema for the keyspace jaeger_v1_dc1 and datacenter dc1
Using template file /cassandra-schema/v001.cql.tmpl with parameters:
    mode = prod
    datacenter = dc1
    keyspace = jaeger_v1_dc1
    replication = {'class': 'NetworkTopologyStrategy', 'dc1': '2' }
    trace_ttl = 172800
    dependencies_ttl = 0
<stdin>:3:OperationTimedOut: errors={'10.46.0.7': 'Client request timeout. See Session.execute[_async](timeout)'}, last_host=10.46.0.7
<stdin>:13:InvalidRequest: Error from server: code=2200 [Invalid query] message="Cannot add type in unknown keyspace jaeger_v1_dc1"
<stdin>:18:InvalidRequest: Error from server: code=2200 [Invalid query] message="Cannot add type in unknown keyspace jaeger_v1_dc1"
<stdin>:24:InvalidRequest: Error from server: code=2200 [Invalid query] message="Cannot add type in unknown keyspace jaeger_v1_dc1"
<stdin>:29:InvalidRequest: Error from server: code=2200 [Invalid query] message="Cannot add type in unknown keyspace jaeger_v1_dc1"
<stdin>:54:ConfigurationException: Keyspace jaeger_v1_dc1 doesn't exist
<stdin>:68:ConfigurationException: Keyspace jaeger_v1_dc1 doesn't exist
<stdin>:83:ConfigurationException: Keyspace jaeger_v1_dc1 doesn't exist
<stdin>:100:ConfigurationException: Keyspace jaeger_v1_dc1 doesn't exist
<stdin>:117:ConfigurationException: Keyspace jaeger_v1_dc1 doesn't exist
<stdin>:136:ConfigurationException: Keyspace jaeger_v1_dc1 doesn't exist
<stdin>:156:ConfigurationException: Keyspace jaeger_v1_dc1 doesn't exist
<stdin>:162:InvalidRequest: Error from server: code=2200 [Invalid query] message="Cannot add type in unknown keyspace jaeger_v1_dc1"
<stdin>:175:ConfigurationException: Keyspace jaeger_v1_dc1 doesn't exist
<stdin>:179:InvalidRequest: Error from server: code=2200 [Invalid query] message="Keyspace jaeger_v1_dc1 does not exist"
```
It seems that it still do not work normal.

I try to recreate jaeger backend. After that, the result give me a surprise that the jaeger backend runs ok:

```
default                    jaeger-collector-216026522-m0tpr       1/1       Running   0          9m        10.46.0.8      ubuntu-k8s-3
default                    jaeger-query-2386461091-lbgr0          1/1       Running   0          9m        10.46.0.9      ubuntu-k8s-3
```

**Comment by jpkrohling** [comment]
The latest template should deploy a `create-schema` job that will reschedule itself upon failure. Could you please confirm you are using a version with this fix? You can get the job's JSON/YAML representation by running something like this:

    $ kubectl get job jaeger-cassandra-schema-job -o yaml

Make sure the `restartPolicy` is set to `OnFailure`. If this is already the case, then I would need more information about your environment, to determine why it hasn't been restarted. 

Also, looks like the `create-schema` job is still using the `0.6` image. I think it should still work, but I'm opening a PR to get this fixed.

About Collector/Query being in the running state even after the failures: I believe this is expected, but if there's something missing (a table, index, ...), it might fail at runtime.

What is *really* strange is that the schema job seemed to work first, and then fail with a `Client request timeout`. This seems to indicate that there was a network failure. Given that the Collector/Query are working, it means that the keyspace `jaeger_v1_dc1` was created.

**Comment by bigwhite** [comment]
@jpkrohling

 The jaeger-cassandra-schema-job info is below:

```
# kubectl get job jaeger-cassandra-schema-job -o yaml
apiVersion: batch/v1
kind: Job
metadata:
  creationTimestamp: 2017-11-07T02:39:39Z
  labels:
    app: jaeger
    jaeger-infra: cassandra-schema-job
  name: jaeger-cassandra-schema-job
  namespace: default
  resourceVersion: "20137088"
  selfLink: /apis/batch/v1/namespaces/default/jobs/jaeger-cassandra-schema-job
  uid: e70e7ef7-c364-11e7-9343-0650cc001a5b
spec:
  activeDeadlineSeconds: 120
  completions: 1
  parallelism: 1
  selector:
    matchLabels:
      controller-uid: e70e7ef7-c364-11e7-9343-0650cc001a5b
  template:
    metadata:
      creationTimestamp: null
      labels:
        controller-uid: e70e7ef7-c364-11e7-9343-0650cc001a5b
        job-name: jaeger-cassandra-schema-job
      name: cassandra-schema
    spec:
      containers:
      - env:
        - name: MODE
          value: prod
        - name: DATACENTER
          value: dc1
        image: jaegertracing/jaeger-cassandra-schema:0.6
        imagePullPolicy: IfNotPresent
        name: jaeger-cassandra-schema
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: OnFailure
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  completionTime: 2017-11-07T02:39:51Z
  conditions:
  - lastProbeTime: 2017-11-07T02:39:51Z
    lastTransitionTime: 2017-11-07T02:39:51Z
    status: "True"
    type: Complete
  startTime: 2017-11-07T02:39:39Z
  succeeded: 1
```

**Comment by jgehrcke** [comment]
@jpkrohling We just had a case where the said job failed and was not rescheduled.

```
Events:
  Type     Reason            Age   From            Message
  ----     ------            ----  ----            -------
  Normal   SuccessfulCreate  56m   job-controller  Created pod: jaeger-cassandra-schema-job-9qt9c
  Normal   SuccessfulDelete  54m   job-controller  Deleted pod: jaeger-cassandra-schema-job-9qt9c
  Warning  DeadlineExceeded  54m   job-controller  Job was active longer than specified deadline
```

```
$ kubectl --namespace jaeger get job jaeger-cassandra-schema-job -o yaml
apiVersion: batch/v1
kind: Job
metadata:
  creationTimestamp: "2020-01-09T14:31:39Z"
  labels:
    app: jaeger
    app.kubernetes.io/component: cronjob-cassandra-schema
    app.kubernetes.io/instance: jaeger
    app.kubernetes.io/managed-by: jaeger-operator
    app.kubernetes.io/name: jaeger-cassandra-schema-job
    app.kubernetes.io/part-of: jaeger
  name: jaeger-cassandra-schema-job
  namespace: jaeger
  ownerReferences:
  - apiVersion: jaegertracing.io/v1
    controller: true
    kind: Jaeger
    name: jaeger
    uid: a32328e8-32ec-11ea-988f-42010aa8011c
  resourceVersion: "3943"
  selfLink: /apis/batch/v1/namespaces/jaeger/jobs/jaeger-cassandra-schema-job
  uid: bf85d3d6-32ec-11ea-988f-42010aa8011c
spec:
  activeDeadlineSeconds: 120
  backoffLimit: 6
  completions: 1
  parallelism: 1
  selector:
    matchLabels:
      controller-uid: bf85d3d6-32ec-11ea-988f-42010aa8011c
  template:
    metadata:
      annotations:
        linkerd.io/inject: disabled
        prometheus.io/scrape: "false"
        sidecar.istio.io/inject: "false"
      creationTimestamp: null
      labels:
        controller-uid: bf85d3d6-32ec-11ea-988f-42010aa8011c
        job-name: jaeger-cassandra-schema-job
    spec:
      containers:
      - env:
        - name: CQLSH_HOST
          value: cassandra-cassandra-nodes.cassandra.svc.cluster.local
        - name: MODE
          value: prod
        - name: DATACENTER
          value: cassandra
        - name: CASSANDRA_USERNAME
        - name: CASSANDRA_PASSWORD
        image: jaegertracing/jaeger-cassandra-schema:1.16.0
        imagePullPolicy: IfNotPresent
        name: jaeger-cassandra-schema
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: OnFailure
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  conditions:
  - lastProbeTime: "2020-01-09T14:33:39Z"
    lastTransitionTime: "2020-01-09T14:33:39Z"
    message: Job was active longer than specified deadline
    reason: DeadlineExceeded
    status: "True"
    type: Failed
  failed: 1
  startTime: "2020-01-09T14:31:39Z"
```

No logs to share yet from the failed job, sorry, they might be gone.
But in that case we certainly want to reschedule the job, too, right?

**Comment by jpkrohling** [comment]
The schema job is idempotent, so, it can be safely rescheduled.

**Comment by jgehrcke** [comment]
> The schema job is idempotent, so, it can be safely rescheduled.

Thanks for the quick reply @jpkrohling -- yes, we did that manually. But is there something we can do in this project here to make this happen automatically? Why did was `restartPolicy: OnFailure` not sufficient in this case? :-).

**Comment by jpkrohling** [comment]
> Why was `restartPolicy: OnFailure` not sufficient in this case? :-).

I guess that's indeed the key question. I would *guess* that `OnFailure` means when the container exits with a non-zero return code, which isn't the case here. A quick look at the docs yields no confirmation on that, though. If you'd be willing to help, could you try to confirm this theory?

**Comment by jgehrcke** [comment]
Still rather new to Kubernetes, but I was reading up on the topic.

The following resources show that the topic of job termination underwent quite some evolution: https://github.com/kubernetes/kubernetes/issues/39578
https://github.com/kubernetes/kubernetes/issues/30243
https://github.com/kubernetes/enhancements/issues/298
https://github.com/kubernetes/community/pull/583/files
https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/

I now understood a few key things I believe:

- A job does not have a `RestartPolicy`.
- The `RestartPolicy` that we talk about in the context of a job actually is the corresponding pod's `RestartPolicy`.
- Hence, a failed job is never restarted magically by some `RestartPolicy`!

That explains the observations above where the system got into a permanent error state that required human intervention.

But... how can a job permanently fail? A job is meant to drive something to completion, no matter what! Usually.

The "no matter what" can be limited via mainly two mechanisms: a maximum pod restart/retry count (`.spec.backoffLimit`) or -- interestingly -- via a global deadline which applies to the entire duration of the job, spanning potentially many pod restarts.

Those two mechanisms I think should be applied rather carefully because they contradict the common reason for why one chooses to use a job in the first place: a job is for "driving something to completion, no matter what" and now we have two brutal methods to abort this process, w/o an automatic strategy handling the 'failed' case.

I think that the main reason to apply either job abortion mechanism is to handle a **critical** problem that cannot self-heal over time, i.e. a problem that requires human intervention, such as component misconfiguration.

In our case here I think we just want to keep trying.

In conclusion I believe that `job/jaeger-cassandra-schema-job` should not have a deadline applied. It should simply re-schedule the corresponding pod until _that_ succeeds, which is the reason for why we have chosen to use a `job` for that in the first place, right? :-)

That is, I think we want to remove `spec.activeDeadlineSeconds`: https://github.com/jaegertracing/jaeger-kubernetes/blob/39dc1369228952164bcdde2350ebc91cf00e36b0/production/cassandra.yml#L125

**Comment by jgehrcke** [comment]
Worth considering:

- With job `.spec.template.spec.restartPolicy = "OnFailure"` the Pod stays on the node, but the container is re-run if schema creation fails.
- With job `.spec.template.spec.restartPolicy = "Never"` if schema creation fails then the container crashes the entire pod and then the Job controller starts a new Pod.

To me it seems the latter might even be more robust in this case here (is more likely to heal various kinds of transient problems).

**Comment by jpkrohling** [comment]
Thanks for your detective work, @jgehrcke! 

> I think that the main reason to apply either job abortion mechanism is to handle a critical problem that cannot self-heal over time, i.e. a problem that requires human intervention, such as component misconfiguration.

Agree, but wouldn't you say, though, that a job that has been running for 24h is in a critical state and needs to be killed? It should certainly be retried, but how do we give up on a specific execution and start a new one? If we just remove `activeDeadlineSeconds`, wouldn't it just hang indefinitely in case it faces a deadlock?

**Comment by jgehrcke** [comment]
:male_detective: :-)

> It should certainly be retried, but how do we give up on a specific execution and start a new one? If we just remove activeDeadlineSeconds, wouldn't it just hang indefinitely in case it faces a deadlock?

Looks like my detective work pays off here, too.. I am really brand new to Kubernetes, but I have found this in the docs for jobs:

> Note that both the Job spec and the [Pod template spec](/docs/concepts/workloads/pods/init-containers/#detailed-behavior) within the Job have an `activeDeadlineSeconds` field. Ensure that you set this field at the proper level.

Sounds like we want to set a deadline on the *pod* itself (and not on the job) to make sure that things are actually retried (whereas the job itself manages the retries).

If you feel better doing so I can try adding this to the PR, or maybe we want to treat that separately?

**Comment by jpkrohling** [comment]
> Sounds like we want to set a deadline on the pod itself (and not on the job) to make sure that the individual attempt does not take longer than T (even if bad code runs into a deadlock), enabling the job to actually retry at a minimum frequency 1/T.

Just to confirm my understanding then: are you saying that if `activeDeadlineSeconds` is set at the `PodSpec` level, the pod will get killed and a new pod will be created, without failing the whole job? That sounds exactly what we need, then.

**Comment by jgehrcke** [comment]
> are you saying that if activeDeadlineSeconds is set at the PodSpec level, the pod will get killed and a new pod will be created, without failing the whole job? That sounds exactly what we need, then.

Yes.

Now it all makes a lot of sense. "Run to completion" seems to be the main purpose of jobs... "Run to Completion" is in the title of https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion

How is "Run to Completion" achieved? It's a little buried in the docs, but I think the key mechanism really is **"When a Pod fails, then the Job controller starts a new Pod"**. Simple, right? :) This also justifies why a _failed job_ itself is a bit of a delicate, weird state, contradicting the idea that a job cannot fail, but runs until it completes. Given that it makes perfect sense that there is no magic automatically handling a _failed_ job. A _failed_ job is an unexpected, critical error.

**Comment by jpkrohling** [comment]
It seems appropriate then to have at both places: in the pod with a 2-minute deadline, and on the job with a really high value (1d?). We certainly don't want to keep retrying executions if we can't get a single clean completion in a whole day, right?

**Comment by jgehrcke** [comment]
> It seems appropriate then to have at both places: in the pod with a 2-minute deadline, and on the job with a really high value (1d?).

I know where you come from. But the exact motivation behind such a high timeout value is a little hard to put into words. And that is a sign for unnecessary complexity.

> We certainly don't want to keep retrying executions if we can't get a single clean completion in a whole day, right?

I am not sure about that. Let's say after a day the job did not succeed. This could manifest in one of two ways:

1) the job is in failed state because of `activeDeadlineSeconds` set to 1 day 
2) the job did simply not yet succeed

First of all, both cases require **human** intervention (we ended up in this state because there is no sane automatic treatment). Secondly, that intervention _will_ happen regardless of whether it is (1) or (2).

If I was the person to handle this case I would prefer (2). Because all I need to do is look at the logs, find the cause, and maybe if it is a permanent dependency error I will fix _that_ and don't have to touch the job at all (the next run would simply succeed). If it turns out that indeed the job is flawed I can decide to delete the job, and restart it after modification.

**Comment by jgehrcke** [comment]
> In other words: what's the advantage of stopping the retry loop?

I think one point of view is: it's nice to at some point see a clear, explicit *failed* somewhere -- so that its' easy to discover which part of the system does not behave as expected.

I choose to not care too much :) Happy to add a short `activeDeadlineSeconds` for the pod, and a long `activeDeadlineSeconds` for the job.

**Comment by jpkrohling** [comment]
> In other words: what's the advantage of stopping the retry loop?

Having an upper bound on what's acceptable helps people not familiar with Jaeger to determine whether something is "still working", or if it's broken. If a schema job can't finish after 1 day worth of retries, then something is definitely wrong.

**Comment by jgehrcke** [comment]
> Happy to add a short activeDeadlineSeconds for the pod, and a long activeDeadlineSeconds for the job.

I have updated #125 with that!
